src/repr/src/row.rs

```diff
@@ -737,6 +832,78 @@ where
     data.extend_from_slice(&u32::to_le_bytes(time.nanosecond()));
 }
 
+/// Returns an i64 representing a `NaiveDateTime`, if
+/// said i64 can be round-tripped back to a `NaiveDateTime`.
+///
+/// The only exotic NDTs for which this can't happen are those that
+/// are hundreds of years in the future or past, or those that
+/// represent a leap second. (Note that Materialize does not support
+/// leap seconds, but this module does).
+// This function is inspired by `NaiveDateTime::timestamp_nanos`,
+// with extra checking.
+fn checked_timestamp_nanos(dt: NaiveDateTime) -> Option<i64> {
+    let subsec_nanos = dt.timestamp_subsec_nanos();
+    if subsec_nanos >= 1_000_000_000 {
+        return None;
+    }
+    let as_ns = dt.timestamp().checked_mul(1_000_000_000)?;
+    as_ns.checked_add(i64::from(subsec_nanos))
+}
+
+#[inline(always)]
+fn min_bytes_signed<T>(i: T) -> usize
+where
+    T: Into<i64>,
+{
+    let i: i64 = i.into();
+
+    // To fit in n bytes, we require that
+    // everything but the leading sign bits fits in n*8 - 1
+    // bits. Why the -1? Because we still need one sign bit
+    // in the encoded data.
+    //
+    // We can improve this, at the cost of doubling the number of extra tags,
+    // by encoding whether the number is positive or negative in the tag.
+    // We could then replace all instances of `7F` with `FF` below.
+    if i == 0 || i.is_positive() {
```

Comment:
IIUC we're trying to determine how many bytes are needed to encode a value? Which I think can be done more succinctly with:
```
let extra_bytes = i.leading_zeros().saturating_sub(1) / 8;
let needed_bytes = std::mem::size_of::<i64>() - extra_bytes;
```

Reply:
That's actually what I initially had, check out the commit "Change how we calculate width" which switched to the nested if statement. (Except, using `leading_ones` in the case of negatives, of course).

I changed it because I suspected the count leading zeros instruction is a bit expensive on x86.

But that might be a premature micro-optimization; maybe I should change it back... it doesn't seem to have made any difference in feature benchmarks.

Reply:
> That's actually what I initially had, check out the commit "Change how we calculate width" which switched to the nested if statement. (Except, using leading_ones in the case of negatives, of course).

Nice! 

_nerd snipe incoming:_ According to these [tables](https://www.agner.org/optimize/instruction_tables.pdf) the instruction used for counting leading zeros, `LZCNT`(which is what Rust/LLVM should use), is no more expensive then a `JMP`, so I don't think counting leading zeros should be any worse. There's other things to consider like speculative execution and binary size, which should also favor counting leading zeros, but this is 1,000% a micro optimization, so I'd go with whatever you think is most clear 😊

Reply:
I agree. I will change it back.